lib/Transforms/IRUtils.cpp

```diff
@@ -0,0 +1,249 @@
+    assert(canSplitBlockAt(block, op));
+
+    // Inline all scopes surrounding `op`.
+    block->walk<mlir::WalkOrder::PostOrder>(
+        [&](P4HIR::ScopeOp scopeOp) { inlineScope(rewriter, scopeOp); });
+
+    // Split resulting block.
+    mlir::Block *before = op->getBlock();
+    mlir::Block *middle = rewriter.splitBlock(before, op->getIterator());
+    mlir::Block *after = rewriter.splitBlock(middle, ++op->getIterator());
+    return {before, middle, after};
+}
+
+namespace {
+
+using BlockSet = llvm::SmallPtrSet<mlir::Block *, 4>;
+
+// Returns blocks that contain uses of val and are not ancestors of `block`.
+BlockSet getEscapingBlocks(mlir::Block *block, mlir::Value val) {
+    BlockSet escapingBlocks;
+    for (mlir::Operation *user : val.getUsers())
+        if (user->getBlock() != block && !block->findAncestorOpInBlock(*user))
+            escapingBlocks.insert(user->getBlock());
+    return escapingBlocks;
+}
+
+// Promote `val` to a local variable.
+// The rewriter's insertion point is the location where the new variable is created.
+void promoteValToVar(mlir::RewriterBase &rewriter, mlir::Value val, const BlockSet &escapingBlocks,
+                     llvm::StringRef varName = "promoted_local") {
+    // Create new variable to hold `val`.
+    auto newVar = rewriter.create<P4HIR::VariableOp>(
+        val.getLoc(), P4HIR::ReferenceType::get(val.getType()), varName);
+
+    mlir::OpBuilder::InsertionGuard guard(rewriter);
+
+    // Insert a new read in all escaping blocks and replace uses of `val` in that block.
+    for (mlir::Block *block : escapingBlocks) {
+        rewriter.setInsertionPointToStart(block);
+        auto newVal = rewriter.create<P4HIR::ReadOp>(val.getLoc(), newVar);
+        rewriter.replaceUsesWithIf(val, newVal, [block](mlir::OpOperand &use) {
+            return use.getOwner()->getBlock() == block;
+        });
+    }
+
+    // Assign the new variable after `val`'s definition.
+    rewriter.setInsertionPointAfterValue(val);
+    rewriter.create<P4HIR::AssignOp>(val.getLoc(), val, newVar);
+}
+
+// Replace uses of `op` with a copy in all escaping blocks.
+void copyOp(mlir::RewriterBase &rewriter, mlir::Operation *op, const BlockSet &escapingBlocks) {
+    mlir::OpBuilder::InsertionGuard guard(rewriter);
+    for (mlir::Block *block : escapingBlocks) {
+        rewriter.setInsertionPointToStart(block);
+        auto newOp = rewriter.clone(*op);
+        rewriter.replaceOpUsesWithinBlock(op, newOp->getResults(), block);
+    }
+}
+
+};  // namespace
+
+void IRUtils::adjustBlockUses(mlir::RewriterBase &rewriter, mlir::Block *block) {
+    // Iterate in reverse so we process uses before their definition in this block.
+    // This is needed to properly handle operands after `copyOp` calls.
+    for (mlir::Operation &op : llvm::make_early_inc_range(llvm::reverse(*block))) {
+        if (mlir::isa<P4HIR::VariableOp>(op)) {
+            // Promote variables to a parser locals.
```

Comment:
```suggestion
            // Move variables to top-level. Rewriter position is expected to point where we can create such locals (e.g. parser or control locals
```

Note that for control locals we'd need to be a bit more elaborated as actions are isolated from above, so they are exported via "symbol references"